### ISSUE TITLE
Possible to trigger based on USB  C port?

### ISSUE THREAD

**Issue by mpoisot** [jakehilborn/displayplacer#45]
I'm trying to figure out a way to automagically use one of two monitor arrangements based on which USB C port I plug my external monitor into. My goal is that when I plug a specific monitor into one of the right two ports, arrange my laptop on the left and the external monitor on the right. When I plug into the left two ports, do the opposite, laptop on the right and external screen on the left.

Can this project do this? I have a feeling I'll need to find another tool to detect the USB event, then trigger this project to rearrange. Thanks for any help.

**Comment by jakehilborn** [comment]
Do some experimenting with using contextual ids instead of persistent ids. You'll need to look at the full output of `displayplacer list` to find the contextual ids. These ids change depending on which port your cable is plugged into. If the contextual ids are indeed different, running both versions of your command with the or operator could work.

`displayplacer "id:1111111111 res:1440x900 scaling:on origin:(0,0) degree:0" || `displayplacer "id:2222222222 res:1440x900 scaling:on origin:(0,0) degree:0"``

**Comment by mpoisot** [comment]
Thanks, that worked!

The contextual ID for the external screen is one value when plugged into the right 2 USB ports, and a different value when plugged into the left 2 ports. I'm a little surprised it doesn't change front vs back ports, but fortunately for my scenario that isn't important.

It looks like displayplacer still modifies things for the screen with the valid ID, even when the other ID is invalid, resulting in a double flash waiting for both settings to get applied. Since my end goal is to trigger then when the display gets plugged in and things already flash and take a few seconds, that won't really matter. Still, it seems like ideally it would fail hard if something isn't right, especially if the shell return code is nonzero.

My next step is to find a way to trigger this script when a USB device is plugged in. I found [Event Scripts](https://www.mousedown.net/software/EventScripts.html) but unfortunately it's not free.

Thanks again for making this useful tool.

**Comment by mpoisot** [comment]
Just to follow up for anyone trying something similar. I paid $7 for Event Scripts, and in combination with DisplayPlacer I got everything working exactly how I wanted.

EventScripts has a GUI to add your custom script, then set the trigger from a list of system events. I found both "USB device attached" and "Spaces Changed" work in terms of detecting that my monitor is connected (via an Anker USB C dock). The problem is that both events are trigger too often for the simple command line @jakehilborn suggested (`displayplacer [config1] || displayplacer [config2]`), because that approach actually sets something every time my script is invoked, which causes a flicker.

When my Anker USB C dock is connected, it registers as 6 different USB devices: ethernet, SD card reader, etc etc. With EventScripts triggering on USB Added events, this resulted in my displayplacer script getting called 6 times in a row, and sometimes it took a good 15 seconds of screen flickering before things settled down. Yikes. And I would get a flicker for any other USB stuff I might connect, unrelated to the monitor. No bueno.

Next I tried triggering EventScripts based on Spaces Changed. This does result in just one event when I connect the monitor. But it also triggers every time I move between spaces - which I do A LOT. These events queue up and result in absurd amounts of screen flickering. Grrr.

What I needed was for my script to do nothing when the screen is already arranged correctly. I looked at the output of `displayplacer list` and realized I could grep a few things to make my script a little smarter. I put the final script [into a gist](https://gist.github.com/mpoisot/5dc29d3aeb2221a8c344240630453bb2).
